File: DSA/DSA_450/Python/10_stack_and_queues/018_Longest_Valid_Parentheses.py
"""
Problem: Length of the Longest Valid Parentheses Substring
URL: https://practice.geeksforgeeks.org/problems/valid-substring0624/1

Problem Statement:
Find length of the longest valid (well-formed) parentheses substring.

Sample Input/Output:
Input: "(()"
Output: 2
Input: ")()())"
Output: 4
Input: "((()()()()(((())"
Output: 8
"""


class Solution:
    def Longest_Valid_Parentheses_Stack(self, s):
        """
        Find longest valid parentheses using stack.
        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        st = [-1]
        maxLen = 0
        for i in range(len(s)):
            if s[i] == '(':
                st.append(i)
            else:
                st.pop()
                if not st:
                    st.append(i)
                else:
                    maxLen = max(maxLen, i - st[-1])
        return maxLen

    def Longest_Valid_Parentheses_Two_Pass(self, s):
        """
        Find longest valid parentheses using two pass approach.
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        left = right = maxLen = 0
        for i in range(len(s)):
            if s[i] == '(':
                left += 1
            else:
                right += 1
            if left == right:
                maxLen = max(maxLen, 2 * right)
            elif right > left:
                left = right = 0
        
        left = right = 0
        for i in range(len(s) - 1, -1, -1):
            if s[i] == '(':
                left += 1
            else:
                right += 1
            if left == right:
                maxLen = max(maxLen, 2 * left)
            elif left > right:
                left = right = 0
        return maxLen


def Test_Longest_Valid_Parentheses():
    solution = Solution()
    
    print("=== Stack Approach ===")
    print(f'Input: "(()"')
    print(f"Output: {solution.Longest_Valid_Parentheses_Stack('(()')}")
    
    print(f'\nInput: ")()())"')
    print(f"Output: {solution.Longest_Valid_Parentheses_Stack(')()())')}")
    
    print(f'\nInput: "((()()()()(((())"')
    print(f"Output: {solution.Longest_Valid_Parentheses_Stack('((()()()()(((())')}")
    
    print(f'\nInput: ""')
    print(f"Output: {solution.Longest_Valid_Parentheses_Stack('')}")
    
    print(f'\nInput: "()(()"')
    print(f"Output: {solution.Longest_Valid_Parentheses_Stack('()(()')}")
    
    print("\n=== Two-Pass Approach ===")
    print(f'Input: "(()"')
    print(f"Output: {solution.Longest_Valid_Parentheses_Two_Pass('(()')}")
    
    print(f'\nInput: ")()())"')
    print(f"Output: {solution.Longest_Valid_Parentheses_Two_Pass(')()())')}")
    
    print(f'\nInput: "((()()()()(((())"')
    print(f"Output: {solution.Longest_Valid_Parentheses_Two_Pass('((()()()()(((())')}")


if __name__ == "__main__":
    Test_Longest_Valid_Parentheses()

File: DSA/DSA_450/10_stack_and_queues/018_Longest_Valid_Parentheses.cpp
/*
Problem: Length of the Longest Valid Parentheses Substring
URL: https://practice.geeksforgeeks.org/problems/valid-substring0624/1

Problem Statement:
Find length of the longest valid (well-formed) parentheses substring.

Sample Input/Output:
Input: "(()"
Output: 2
Input: ")()())"
Output: 4
Input: "((()()()()(((())"
Output: 8
*/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int Longest_Valid_Parentheses_Stack(string s) {
        stack<int> st;
        st.push(-1);
        int maxLen = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s[i] == '(') {
                st.push(i);
            } else {
                st.pop();
                if (st.empty()) {
                    st.push(i);
                } else {
                    maxLen = max(maxLen, i - st.top());
                }
            }
        }
        return maxLen;
    }

    int Longest_Valid_Parentheses_Two_Pass(string s) {
        int left = 0, right = 0, maxLen = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s[i] == '(') left++;
            else right++;
            if (left == right) maxLen = max(maxLen, 2 * right);
            else if (right > left) left = right = 0;
        }
        left = right = 0;
        for (int i = s.length() - 1; i >= 0; i--) {
            if (s[i] == '(') left++;
            else right++;
            if (left == right) maxLen = max(maxLen, 2 * left);
            else if (left > right) left = right = 0;
        }
        return maxLen;
    }
};

void Test_Longest_Valid_Parentheses() {
    Solution solution;
    
    cout << "=== Stack Approach ===" << endl;
    cout << "Input: \"(()\"" << endl;
    cout << "Output: " << solution.Longest_Valid_Parentheses_Stack("(()") << endl;
    
    cout << "\nInput: \")()())\"" << endl;
    cout << "Output: " << solution.Longest_Valid_Parentheses_Stack(")()())") << endl;
    
    cout << "\nInput: \"((()()()()(((())" << endl;
    cout << "Output: " << solution.Longest_Valid_Parentheses_Stack("((()()()()(((())") << endl;
    
    cout << "\nInput: \"\"" << endl;
    cout << "Output: " << solution.Longest_Valid_Parentheses_Stack("") << endl;
    
    cout << "\nInput: \"()(()\"" << endl;
    cout << "Output: " << solution.Longest_Valid_Parentheses_Stack("()(()") << endl;
    
    cout << "\n=== Two-Pass Approach ===" << endl;
    cout << "Input: \"(()\"" << endl;
    cout << "Output: " << solution.Longest_Valid_Parentheses_Two_Pass("(()") << endl;
    
    cout << "\nInput: \")()())\"" << endl;
    cout << "Output: " << solution.Longest_Valid_Parentheses_Two_Pass(")()())") << endl;
    
    cout << "\nInput: \"((()()()()(((())" << endl;
    cout << "Output: " << solution.Longest_Valid_Parentheses_Two_Pass("((()()()()(((())") << endl;
}

int main() {
    Test_Longest_Valid_Parentheses();
    return 0;
}
